File: Funcional/Info.hs
module Info where

import System.IO.Unsafe
import System.Process
import System.Info
import Control.Concurrent
import Util

clrScr = if os == "mingw32"
            then system "cls" 
         else system "clear"

listaLoading = [
    "arquivos/loading1.txt",
    "arquivos/loading2.txt",
    "arquivos/loading3.txt",
    "arquivos/loading4.txt"]

putLogoInicial :: IO()
putLogoInicial = do
    carregaLogoInicial listaLoading "            INICIANDO O"

carregaLogoInicial :: [String] -> String -> IO()
carregaLogoInicial [] _ = threadDelay 200000
carregaLogoInicial (x:xs) msg = do {formataLogoInicial x msg; carregaLogoInicial xs msg}

formataLogoInicial :: String -> String -> IO()
formataLogoInicial x msg = do
    clrScr
    putStrLn("---")
    putStrLn(msg)
    putStrLn(unsafeDupablePerformIO(readFile x))
    threadDelay 200000

putLogo :: IO()
putLogo = do
    clrScr
    putStrLn(unsafeDupablePerformIO(readFile "arquivos/logo.txt"))

putMsgSaida :: IO()
putMsgSaida = do
    carregaLogoInicial (reverse listaLoading) "            SAINDO DO"
    clrScr
    putStrLn(unsafeDupablePerformIO(readFile "arquivos/corta.txt"))
    threadDelay 1000000
    clrScr
    putStr("")

putMsgOpcoesMenuInicial :: IO()
putMsgOpcoesMenuInicial = do
    putLogo
    putStrLn("       ----MENU PRINCIPAL----")
    putStrLn("\nComo deseja prosseguir?")
    putStrLn("[1] Login como cliente")
    putStrLn("[2] Login como administrador")
    putStrLn("[3] Cadastro de usuário")
    putStrLn("[4] Sair\n")

putMsgOpcoesMenuCliente :: String -> IO()
putMsgOpcoesMenuCliente nome = do
    putLogo
    putStrLn("       --LUZ, CÂMERA, AÇÃO!--")
    putStrLn("\nOlá, " ++ nome ++ "!")
    putStrLn("\nComo deseja prosseguir?")
    putStrLn("[1] Listar filmes")
    putStrLn("[2] Fazer locação")
    putStrLn("[3] Recomendação da locadora")
    putStrLn("[4] Devolução da locação")
    putStrLn("[5] Voltar ao menu principal\n")

putMsgOpcoesMenuAdmin :: IO()
putMsgOpcoesMenuAdmin = do
    putLogo
    putStrLn("       -----ADMINISTRADOR----")
    putStrLn("\nOlá, Administrador!")
    putStrLn("\nComo deseja prosseguir?")
    putStrLn("[1] Cadastrar filme")
    putStrLn("[2] Exibir histórico de locações")
    putStrLn("[3] Gerenciar estoque")
    putStrLn("[4] Voltar ao menu principal\n")

putMsgOpcoesExibirLocacoes :: IO()
putMsgOpcoesExibirLocacoes = do
    putLogo
    putStrLn("       -----ADMINISTRADOR----")
    putStrLn("\nComo deseja prosseguir?")
    putStrLn("[1] Exibir histórico geral")
    putStrLn("[2] Exibir histórico do cliente")
    putStrLn("[3] Exibir locações em andamento")
    putStrLn("[4] Voltar\n")

putMsgOpcoesGerenciarEstoque :: IO()
putMsgOpcoesGerenciarEstoque = do
    putLogo
    putStrLn("       -----ADMINISTRADOR----")
    putStrLn("\nComo deseja prosseguir?")
    putStrLn("[1] Adicionar estoque ao filme")
    putStrLn("[2] Verificar disponibilidade de filmes")
    putStrLn("[3] Voltar ao menu inicial\n")

putMsgOpcaoInvalida :: IO()
putMsgOpcaoInvalida = do
    putStrLn("\nErro: opção inválida!")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgCpfInvalido :: IO()
putMsgCpfInvalido = do
    putStrLn("\nErro: CPF inválido!")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgUserInvalido :: IO()
putMsgUserInvalido = do
    putStrLn("\nErro: usuário não cadastrado!")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgIdInvalido :: IO()
putMsgIdInvalido = do
    putStrLn("\nErro: id inválido!")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgGeneroInvalido :: IO()
putMsgGeneroInvalido = do
   putStrLn("\nErro: gênero inválido!")
   putMsgTeclaEnter
   opcao <- Util.lerEntradaString
   putStr("")

putMsgQuantidadeInvalida :: IO()
putMsgQuantidadeInvalida = do
   putStrLn("\nErro: quantidade inválida!")
   putMsgTeclaEnter
   opcao <- Util.lerEntradaString
   putStr("")

putMsgFilmeNaoCadastrado :: IO()
putMsgFilmeNaoCadastrado  = do
   putStrLn("\nErro: filme não cadastrado!")
   putMsgTeclaEnter
   opcao <- Util.lerEntradaString
   putStr("")

putMsgFilmeNaoDisponivel::IO()
putMsgFilmeNaoDisponivel = do
   putStrLn("\nErro: filme indisponível!")
   putMsgTeclaEnter
   opcao <- Util.lerEntradaString
   putStr("")

putMsgLoginCliente :: IO() 
putMsgLoginCliente = do
    putLogo
    putStrLn("       -----LOGIN CLIENTE----\n")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")
    putStrLn("\nOlá, cinéfilo! :)\n")

    putStrLn("Informe o seu CPF para continuar.")
    putStrLn("\nCPF (apenas números): ")


putMsgCadastroNome :: IO()
putMsgCadastroNome = do
    putLogo
    putStrLn("       -------CADASTRO-------")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")   
    putStrLn("\nNome do usuário: ")

putMsgCadastroCpf :: IO()
putMsgCadastroCpf = do
    putStrLn("--- \nCPF (apenas números): ")

putMsgCadastroTelefone :: IO()
putMsgCadastroTelefone = do
    putStrLn("--- \nTelefone (DDD + Número): ")

putMsgCadastroEndereco :: IO()
putMsgCadastroEndereco = do
    putStrLn("--- \nEndereço: ")

putMsgResumoCadastroUsuario :: String -> IO()
putMsgResumoCadastroUsuario cliente = do
    putLogo
    putStrLn("       -------CADASTRO-------")
    putStrLn("\n-----\n")
    putStrLn(cliente)
    putStrLn("\n-----")
    putMsgTeclaEnter

putMsgCadastroFilmeTitulo :: IO()
putMsgCadastroFilmeTitulo = do
    putLogo
    putStrLn("       -------CADASTRO-------")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")
    putStr ("")
    putStrLn("\nTítulo: ")

putMsgCadastroFilmeGenero :: IO()
putMsgCadastroFilmeGenero = do
    putStrLn("Gênero: ")

putMsgCadastroFilmeDiretor :: IO()
putMsgCadastroFilmeDiretor = do
    putStrLn("Diretor(a): ")

putMsgCadastroFilmeData :: IO()
putMsgCadastroFilmeData = do
    putStrLn("Data de lançamento (dd/mm/aaaa): ")

putMsgFilmeQuantidade :: IO()
putMsgFilmeQuantidade = do
    putStrLn("Quantidade: ")

putMsgResumoCadastroFilme :: String -> IO()
putMsgResumoCadastroFilme filme = do
    putLogo
    putStrLn("       -------CADASTRO-------")
    putStrLn("\n-----\n")
    putStrLn(filme)
    putStrLn("\n-----")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgFilmeIdentificador :: IO()
putMsgFilmeIdentificador = do
    putStrLn("\nID do filme: ")

putMsgLocacaoFilme :: IO()
putMsgLocacaoFilme = do
    putLogo
    putStrLn("       -VAI UM FILMINHO AI?-")
    putStrLn("\nOBS: Para verificar a lista de filmes basta digitar 'L'!")
    putStrLn ("     Para voltar ao menu basta digitar 'S'!\n")
    putStrLn("\nID do filme: ")

putMsgDataLocacao :: IO()
putMsgDataLocacao = do
   putStrLn("Insira a data da locação (dd/mm/aaaa): ")


putMsgLocaFilme :: String -> String -> IO()
putMsgLocaFilme nomeFilme qtd = do
    putStrLn("\nJá pode ir preparando a pipoca...")
    putStrLn("Filme " ++ nomeFilme ++ " alugado com sucesso!")
    putStrLn("Agora só temos: " ++ qtd ++ " unidade(s) disponivel(is).")
    putStrLn("---")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

putMsgHistoricoLocacoes :: IO()
putMsgHistoricoLocacoes = do
    putLogo
    putStrLn("       --HISTÓRICO LOCAÇÕES--")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")

putMsgHistoricoLocacoesCpf :: IO()
putMsgHistoricoLocacoesCpf = do
    putStrLn("\nDigite o CPF do cliente: ")

putMsgHistoricoLocacoesAndamento :: IO()
putMsgHistoricoLocacoesAndamento = do
    putLogo
    putStrLn("       --LOCAÇÕES ANDAMENTO--")

putMsgEstoqueFilmes :: IO()
putMsgEstoqueFilmes = do
    putLogo
    putStrLn("       ---ADICIONA ESTOQUE---")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")

putMsgDisponibilidadeFilmes :: IO()
putMsgDisponibilidadeFilmes = do
    putLogo
    putStrLn("       ----DISPONIBILIDADE---")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")

putMsgRecomendacaoGenero :: IO()
putMsgRecomendacaoGenero = do
    putLogo
    putStrLn("       -----RECOMENDAÇÃO-----")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")   
    putStrLn("\nInsira o gênero: ")

putMsgRecomendacao :: String -> IO()
putMsgRecomendacao filme = do
    putLogo
    putStrLn("       -HMM VEJAMOS, JÁ SEI!-")
    putStrLn ("\nOBS: Para voltar ao menu basta digitar 'S'!\n")   
    putStrLn("\nBaseado no seu perfil, recomendamos o seguinte filme:")
    putStrLn("\n"++ filme ++"\n")

putMsgRecomendaLocacao :: IO()
putMsgRecomendaLocacao = do
    putStrLn("Você deseja fazer a locação desse filme? [y/n]")

putMsgListaFilmes :: IO()
putMsgListaFilmes = do
    putLogo
    putStrLn("       ----LISTA DE FILMES---")

putMsgTeclaEnter :: IO()
putMsgTeclaEnter = do
    putStrLn("\nPressione a tecla ENTER para voltar")

putMsgDevolucaoTop :: IO()
putMsgDevolucaoTop = do
    putLogo
    putStrLn("       -------DEVOLUÇÃO------")
    putStrLn("\nUau, já assistiu?!")
    putStrLn("Você tem a(s) seguinte(s) locação(ões) em andamento:")

putMsgDevolucaoBottom :: IO()
putMsgDevolucaoBottom = do
    putStrLn("Qual locação você deseja encerrar?")
    putStrLn("OBS: Para voltar ao menu basta digitar 'S'!")
    putStrLn("\nID da locação: ")

putMsgDevolveFilme :: String -> IO()
putMsgDevolveFilme qtd = do
    putStrLn("\nDevolução realizada, esperamos que tenha gostado!")
    putStrLn("Agora temos: " ++ qtd ++ " unidade(s) disponivel(is).")
    putStrLn("---")
    putMsgTeclaEnter
    opcao <- Util.lerEntradaString
    putStr("")

File: Logico/Info.pl

:- include('Util.pl').

putOpcoesHistorico :-
    put_logo_historico,
    writeln('Como deseja prosseguir?'),
    writeln('[1] Exibir histórico geral'),
    writeln('[2] Exibir histórico do cliente'),
    writeln('[3] Exibir locações em andamento'),
    writeln('[4] Voltar ao menu inicial').

putMsgOpcoesMenuInicial() :-
    put_logo,
    writeln('       ----MENU PRINCIPAL----'),
    writeln('\nComo deseja prosseguir?'),
    writeln('[1] Login como cliente'),
    writeln('[2] Login como administrador'),
    writeln('[3] Cadastro de usuário'),
    writeln('[4] Sair\n').

putMsgOpcoesMenuCliente(Nome) :-
    put_logo,
    writeln('       --LUZ, CÂMERA, AÇÃO!--'),
    string_concat('             \nOlá, ', Nome, Return),
    write(''),
    write(Return),
    writeln('!'),
    writeln('\nComo deseja prosseguir?'),
    writeln('[1] Listar filmes'),
    writeln('[2] Fazer locação'),
    writeln('[3] Recomendação da locadora'),
    writeln('[4] Devolução da locação'),
    writeln('[5] Voltar ao menu principal\n').

putMsgOpcoesMenuAdmin() :- 
    put_logo,
    writeln('       -----ADMINISTRADOR----'),
    writeln('\nOlá, Administrador!'),
    writeln('\nComo deseja prosseguir?'),
    writeln('[1] Cadastrar filme'),
    writeln('[2] Exibir histórico de locações'),
    writeln('[3] Gerenciar estoque'),
    writeln('[4] Voltar ao menu principal\n'), halt().

putMsgOpcoesExibirLocacoes() :-
    put_logo,
    writeln('       -----ADMINISTRADOR----'),
    writeln('\nComo deseja prosseguir?'),
    writeln('[1] Exibir histórico geral'),
    writeln('[2] Exibir histórico do cliente'),
    writeln('[3] Exibir locações em andamento'),
    writeln('[4] Voltar\n'), halt().

putMsgOpcoesGerenciarEstoque() :-
    put_logo,
    writeln('       -----ADMINISTRADOR----'),
    writeln('\nComo deseja prosseguir?'),
    writeln('[1] Adicionar estoque ao filme'),
    writeln('[2] Verificar disponibilidade de filmes'),
    writeln('[3] Voltar ao menu inicial\n'), halt().

putMsgOpcaoInvalida() :- writeln('\nErro: opção inválida!').

putMsgCpfInvalido() :- 
    writeln('\nErro: CPF inválido!'),
    %putMsgTeclaEnter()
    %opcao = Util.lerEntradaString
    write(''), halt().

putMsgUserInvalido() :-
    writeln('\nErro: usuário não cadastrado!'),
    %putMsgTeclaEnter()
    %opcao = Util.lerEntradaString
    write(''), halt().

putMsgIdInvalido() :-
    writeln('\nErro: id inválido!'),
    %putMsgTeclaEnter()
    %opcao = Util.lerEntradaString
    write(''), halt().

putMsgGeneroInvalido() :-
   writeln('\nErro: gênero inválido!'),
   %putMsgTeclaEnter()
   %opcao = Util.lerEntradaString
   write(''), halt().

putMsgQuantidadeInvalida() :-
   writeln('\nErro: quantidade inválida!'),
   %putMsgTeclaEnter()
   %opcao = Util.lerEntradaString
   write(''), halt().

putMsgFilmeNaoCadastrado() :-
   writeln('\nErro: filme não cadastrado!'),
   %putMsgTeclaEnter()
   %opcao = Util.lerEntradaString
   write(''), halt().

putMsgFilmeNaoDisponivel() :-
   writeln('\nErro: filme indisponível!'),
   %putMsgTeclaEnter()
   %opcao = Util.lerEntradaString
   write(''), halt().

putMsgLoginCliente() :-
    put_logo,
    writeln('       -----LOGIN CLIENTE----\n'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'),
    writeln('\nOlá, cinéfilo! :)\n'),

    writeln('Informe o seu CPF para continuar.'),
    writeln('\nCPF (apenas números): '), halt().


putMsgCadastroNome() :-
    put_logo,
    writeln('       -------CADASTRO-------'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'),
    writeln('\nNome do usuário: '), halt().

putMsgCadastroCpf() :-
    writeln('--- \nCPF (apenas números): '), halt().

putMsgCadastroTelefone() :-
    writeln('--- \nTelefone (DDD + Número): '), halt().

putMsgCadastroEndereco() :-
    writeln('--- \nEndereço: '), halt().

putMsgResumoCadastroUsuario(cliente) :-
    put_logo,
    writeln('       -------CADASTRO-------'),
    writeln('\n-----\n'),
    writeln(cliente),
    writeln('\n-----'),

    %putMsgTeclaEnter().
    halt().

putMsgCadastroFilmeTitulo() :-
    put_logo,
    writeln('       -------CADASTRO-------'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'),
    write(''),
    writeln('\nTítulo: '), halt().

putMsgCadastroFilmeGenero() :-
    writeln('Gênero: '), halt().

putMsgCadastroFilmeDiretor() :-
    writeln('Diretor(a): '), halt().

putMsgCadastroFilmeData() :-
    writeln('Data de lançamento (dd/mm/aaaa): '), halt().

putMsgFilmeQuantidade() :-
    writeln('Quantidade: '), halt().

putMsgResumoCadastroFilme(Filme) :-
    %put_logo()
    writeln('       -------CADASTRO-------'),
    writeln('\n-----\n'),
    writeln(Filme),
    writeln('\n-----'),
    %putMsgTeclaEnter().
    %opcao = Util.lerEntradaString,
    write(''), halt().

putMsgFilmeIdentificador() :-
    writeln('\nID do filme: '), halt().

putMsgLocacaoFilme() :-
    put_logo,
    writeln('       -VAI UM FILMINHO AI?-'),
    writeln('\nOBS: Para verificar a lista de filmes basta digitar "L"!'),
    writeln('     Para voltar ao menu basta digitar "S"!\n'),
    writeln('\nID do filme: '), halt().

putMsgDataLocacao() :-
   writeln('Insira a data da locação (dd/mm/aaaa): '), halt().

putMsgLocaFilme(NomeFilme, Qtd) :-
    writeln('\nJá pode ir preparando a pipoca...'),
    writeln(''),
    string_concat('Filme, ', NomeFilme, Return),
    write(Return),
    writeln(' alugado com sucesso!'),
    writeln(''),
    string_concat('Agora só temos: ', Qtd, R),
    write(R),
    writeln(' unidade(s) disponivel(is).'),
    writeln('---'),
    %putMsgTeclaEnter()
    %opcao = Util.lerEntradaString
    write(''), halt().

putMsgHistoricoLocacoes() :-
    put_logo,
    writeln('       --HISTÓRICO LOCAÇÕES--'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'), halt().

putMsgHistoricoLocacoesCpf() :-
    writeln('\nDigite o CPF do cliente: '), halt().

putMsgHistoricoLocacoesAndamento() :-
    put_logo,
    writeln('       --LOCAÇÕES ANDAMENTO--'), halt().

putMsgEstoqueFilmes() :-
    put_logo,
    writeln('       ---ADICIONA ESTOQUE---'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'), halt().

putMsgDisponibilidadeFilmes() :-
    put_logo,
    writeln('       ----DISPONIBILIDADE---'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'), halt().

putMsgRecomendacaoGenero() :-
    put_logo,
    writeln('       -----RECOMENDAÇÃO-----'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'),   
    writeln('\nInsira o gênero: '), halt().

putMsgRecomendacao(NomeFilme) :-
    put_logo,
    writeln('       -HMM VEJAMOS, JÁ SEI!-'),
    writeln('\nOBS: Para voltar ao menu basta digitar "S"!\n'),   
    write('\nBaseado no seu perfil, recomendamos o seguinte filme: '),
    writeln(NomeFilme),
    writeln(''), halt().

putMsgRecomendaLocacao() :-
    writeln('Você deseja fazer a locação desse filme? [y/n]'), halt().

putMsgListaFilmes() :-
    put_logo,
    writeln('       ----LISTA DE FILMES---'), halt().

putMsgTeclaEnter() :-
    writeln('\nPressione a tecla ENTER para voltar'), halt().

putMsgDevolucaoTop() :-
    put_logo,
    writeln('       -------DEVOLUÇÃO------'),
    writeln('\nUau, já assistiu?!'),
    writeln('Você tem a(s) seguinte(s) locação(ões) em andamento:'), halt().

putMsgDevolucaoBottom() :-
    writeln('Qual locação você deseja encerrar?'),
    writeln('OBS: Para voltar ao menu basta digitar "S"!'),
    writeln('\nID da locação: '), halt().

putMsgDevolveFilme(Qtd):-
    writeln('\nDevolução realizada, esperamos que tenha gostado!'),
    writeln(''),
    string_concat('Agora temos: ', Qtd, Return),
    write(Return),
    writeln(' unidade(s) disponivel(is).'),
    write('---'),
    %putMsgTeclaEnter().
    %opcao = Util.lerEntradaString
    write(''), halt().